sklearn/preprocessing/knn_imputation.py

```diff
@@ -0,0 +1,253 @@
+from ..utils import check_array
+from ..utils.validation import check_is_fitted
+from ..utils.validation import FLOAT_DTYPES
+from ..neighbors import NearestNeighbors
+from ..neighbors.base import _get_weights, _check_weights
+__all__ = [
+    'KNNImputer',
+]
+
+
+def _get_mask(X, value_to_mask):
+    """Compute the boolean mask X == missing_values."""
+    if value_to_mask == "NaN" or np.isnan(value_to_mask):
+        return np.isnan(X)
+    else:
+        return X == value_to_mask
+
+
+class KNNImputer(BaseEstimator, TransformerMixin):
+    """Imputation transformer for completing missing values using Nearest
+    Neighbors. Broadly speaking, the imputation is performed using either
+    the weighted or the unweighted mean of the desired number of neighbors.
+
+    Parameters
+    ----------
+    missing_values : integer or "NaN", optional (default="NaN")
+        The placeholder for the missing values. All occurrences of
+        `missing_values` will be imputed. For missing values encoded as np.nan,
+        use the string value "NaN".
+
+    n_neighbors : int, optional (default = 5)
+        Number of neighbors to get.
+
+    weights : str or callable
+        weight function used in prediction.  Possible values:
+
+        - 'uniform' : uniform weights.  All points in each neighborhood
+          are weighted equally.
+        - 'distance' : weight points by the inverse of their distance.
+          in this case, closer neighbors of a query point will have a
+          greater influence than neighbors which are further away.
+        - [callable] : a user-defined function which accepts an
+          array of distances, and returns an array of the same shape
+          containing the weights.
+
+        Uniform weights are used by default.
+
+    metric : string or callable, optional (default = 'masked_euclidean')
+        metric to use for distance computation.
+
+    row_max_missing: float, optional (default = 0.5)
+        The maximum percentage of columns (i.e. features) that can be missing
+        before the sample is excluded from nearest neighbor imputation. It
+        means that such rows will not be considered a potential donor in fit()
+        and in transform() their missing feature values will be imputed to be
+        the column mean for the entire dataset.
+
+    col_max_missing: float, optional (default = 0.8)
+        The maximum percentage of rows (or samples) that can be missing
+        for a given feature beyond which an error is raised.
+
+    copy : boolean, optional (default=True)
+        If True, a copy of X will be created. If False, imputation will
+        be done in-place whenever possible. Note that, if metric is
+        "masked_euclidean" and copy=False then missing_values in the
+        input matrix X will be overwritten with zeros.
+
+    Attributes
+    ----------
+    statistics_ : {tuple}
+        A tuple whose first element is the input dataset used to fit the
```

Comment:
No, don't use a tuple. Just store the means here.

If we need another attribute to store the NearestNeighbors instance, we can do that.